odoo/addons/base/transpiler/transpiler_js.py

```diff
@@ -0,0 +1,393 @@
+        #    (?P<module>        # Capture in a group named module
+        #        \w+            # One or more word character
+        #    )
+        #    /                  # /
+        #    [\w/]*             # A word character or /, zero to many times.
+        #    static             # static
+        #    /                  # /
+        #    (?P<type>          # Capture in a group named type
+        #        src|tests      # src or tests
+        #    )
+        #    /                  # /
+        #    (?P<url>           # Capture in a group named url
+        #        [\w/]*         # A word character or /, zero to many times.
+        #    )
+        #    """, url, re.VERBOSE)
+        if result:
+            d = result.groupdict()
+            url = d.get("url")
+            if (url.split("/")[-1] in ["index.js", "index"]):
+                url = "/".join(url.split("/")[:-1])
+            if d.get("type") == "src":
+                return "@%s/%s" % (d.get('module'), url)
+            else:
+                return "@%s/../tests/%s" % (d.get('module'), url)
+        else:
+            raise ValueError("The js file '%s' must be in the folder '/static/src' or '/static/test'" % url)
+
+    def wrap_with_odoo_define(self):
+        """
+        Wraps the current content (source code) with the odoo.define call.
+        Should logically be called once all other operations have been performed.
+        """
+        self.content = f"odoo.define('{self.module_path}', function (require) {{\
+                \n'use strict';\
+                \nlet __exports = {{}};\
+                \n{self.content}\
+                \nreturn __exports;\
+                \n}});\n"
+
+    def convert_function_and_class_export(self, default=False):
+        """
+        Transpile functions and classes that are being exported.
+
+        before: export function name
+        after: const name = __exports.name = function name
+
+        before: export async function name
+        after: const name = __exports.name = async function name
+
+        before: export class name
+        after: const name = __exports.name = class name
+
+        if default:
+        before: export default function name
+        after: const name = __exports[Symbol.for("default")] = function name
+        (same for the rest)
+
+        :param default: should we convert the default exports as well ?
+        """
+        pattern = r"^(?P<space>\s*)export\s+(?P<type>(async\s+)?function|class)\s+(?P<identifier>\w+)"
+        repl = r"\g<space>const \g<identifier> = __exports.\g<identifier> = \g<type> \g<identifier>"
+        if default:
+            pattern = r"^(?P<space>\s*)export\s+default\s+(?P<type>(async\s+)?function|class)\s+(?P<identifier>\w+)"
+            repl = r"""\g<space>const \g<identifier> = __exports[Symbol.for("default")] = \g<type> \g<identifier>"""
+        p = re.compile(pattern, re.MULTILINE)
+        self.content = p.sub(repl, self.content)
```

Comment:
Not actually useful. `re.compile` is useful at the toplevel in order to keep a handle on the compiled pattern (as a constant), here you can just use `re.sub` that has the same effect.

Reply:
(this applis to all occurrences of this pattern, obviously)

Reply:
Wouldn't it be more efficient to put all the regex compiled in a global variable at the beginning of the file to only compile them once?

Reply:
Yes, you could store the result of each `re.compile` in a constant. Whether it would be more efficient I do not quite know: that's quite a lot of stuff to store, so it depends *how* it's called. If this is only used a few times at the start, then the *regex cache* would handle it fine: the first time it's called it's going to be filled, the next few calls will get the compiled patterns from the cache, then the cached patterns will be evicted during normal odoo runtime (as regex get used elsewhere). As external constants they would essentially never be evicted. Which is better if the regex can be expected to be used essentially continuously.

Reply:
Either would work, really. But that's a consideration if you're really looking for efficiency.